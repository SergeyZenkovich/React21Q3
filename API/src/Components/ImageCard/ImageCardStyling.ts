import styled from "styled-components";
import device from "../../utils/sizes";

export const StyledCard = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 24%;
  @media ${device.mobileS} {
    min-width: 160px;
  }
  @media ${device.laptop} {
    min-width: 275px;
  }
  height: 50%;
  min-height: 200px;
  margin: 15px 5px;
  border: 1px solid #000;
  & img {
    width: 100%;
    height: 100%;
  }
`;
