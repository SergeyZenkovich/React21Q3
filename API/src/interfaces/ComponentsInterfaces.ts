export interface CardInterface {
  cardProblem: string;
  cardDate: string;
  cardPriority: string;
  cardisRequestNeeded: boolean;
}
export interface ImageCardInterface {
  url: string;
}
export interface FetchObject {
  alt_description: string;
  color: string;
  id: string;
  likes: number;
  urls: FetchedURLS;
  user: FetchedUser;
}
export interface URLs {
  urls?: {
    full: string;
    regular: string;
    small: string;
    thumb: string;
  };
}
export interface FetchedURLS {
  full: string;
  regular: string;
  small: string;
  thumb: string;
}
export interface FetchedUser {
  first_name: string;
  last_name: string;
  location: string;
  username: string;
  twitter_username: string;
  instagram_username: string;
  total_photos: number;
  total_likes: number;
}
export interface NavElement {
  page: string;
  link: string;
}
export interface QueryParametersInterface {
  query: string;
  page: number;
  elementsOnPage: number;
  orient: string;
  color: string;
  orderBy: string;
}
export interface IdParameterInterface {
  id: string;
}
export interface RequestResponseInterface {
  results: FetchObject[];
  total_pages: number;
}
