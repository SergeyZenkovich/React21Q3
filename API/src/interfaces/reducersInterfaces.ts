import {
  FetchedURLS,
  FetchedUser,
  FetchObject,
  QueryParametersInterface
} from "./ComponentsInterfaces";

/// home page reducer

export interface HomePageStateInterface {
  totalElementsCount: number;
  elements: FetchObject[];
  isFetching: boolean;
  requestParams: QueryParametersInterface;
}

export enum FetchActionsType {
  TOGGLE_FETCHING = "TOGGLE-LOADING",
  SET_TOTAL_COUNT = "SET-TOTAL-COUNT",
  SET_ELEMENTS = "SET-ELEMENTS",
  SET_PAGE_PARAM = "SET-PAGE-PARAM",
  SET_QUERY_PARAM = "SET-QUERY-PARAM",
  SET_ORIENT_PARAM = "SET-ORIENT-PARAM",
  SET_COLOR_PARAM = "SET-COLOR-PARAM",
  SET_OREDER_BY_PARAM = "SET-ORDER-BY-PARAM",
  SET_ELEMENTS_ON_PAGE_PARAM = "SET-ELEMENTS-ON-PAGE-PARAM",
  SET_PARAMS = "SET-PARAMS"
}

export interface LoadingAction {
  type: FetchActionsType.TOGGLE_FETCHING;
}
export interface SetTotalCountAction {
  type: FetchActionsType.SET_TOTAL_COUNT;
  totalCount: number;
}
export interface SetElementsAction {
  type: FetchActionsType.SET_ELEMENTS;
  elements: FetchObject[];
}
export interface SetPageParamAction {
  type: FetchActionsType.SET_PAGE_PARAM;
  page: number;
}
export interface SetQueryParamAction {
  type: FetchActionsType.SET_QUERY_PARAM;
  query: string;
}
export interface SetOrientParamAction {
  type: FetchActionsType.SET_ORIENT_PARAM;
  orient: string;
}
export interface SetColorParamAction {
  type: FetchActionsType.SET_COLOR_PARAM;
  color: string;
}
export interface SetOrderByParamAction {
  type: FetchActionsType.SET_OREDER_BY_PARAM;
  orderBy: string;
}
export interface SetElementsOnPageParamAction {
  type: FetchActionsType.SET_ELEMENTS_ON_PAGE_PARAM;
  elementsOnPage: number;
}
export interface SetParamsAction {
  type: FetchActionsType.SET_PARAMS;
  params: QueryParametersInterface;
}

export type FetchActions =
  | LoadingAction
  | SetTotalCountAction
  | SetElementsAction
  | SetParamsAction
  | SetPageParamAction
  | SetQueryParamAction
  | SetOrientParamAction
  | SetColorParamAction
  | SetOrderByParamAction
  | SetElementsOnPageParamAction;

/// details page reducer

export interface DetailsPageStateInterface {
  isLoading: boolean;
  element: {
    id: string;
    color: string;
    alt_description: string;
    likes: number;
    urls: FetchedURLS;
    user: FetchedUser;
  };
}

export enum FetchActionsTypeDetails {
  TOGGLE_FETCHING = "TOGGLE-LOADING",
  SET_ELEMENT = "SET-ELEMENT"
}

export interface SetDetailsElementAction {
  type: FetchActionsTypeDetails.SET_ELEMENT;
  element: FetchObject;
}
export interface LoadingDetailsAction {
  type: FetchActionsTypeDetails.TOGGLE_FETCHING;
}

export type DetailsFetchingActions = SetDetailsElementAction | LoadingDetailsAction;
