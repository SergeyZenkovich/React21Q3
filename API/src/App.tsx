import React, {useEffect, useState} from "react";
import {Switch, Route} from "react-router-dom";
import styled from "styled-components";

import HomePage from "./Components/HomePage/HomePage";
import AboutPage from "./Components/AboutPage/AboutPage";
import Header from "./Components/Header/Header";
import ErrorBlock from "./Components/ErrorComponent/ErrorComponent";
import DetailsPage from "./Components/DetailsPage/DetailsPage";
import {FetchObject} from "./interfaces/ComponentsInterfaces";

const MainComponent = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
`;

const App = (): JSX.Element => {
  const [activePage, setActivePage] = useState("Home");
  const [pageElements, setPageElements] = useState([] as FetchObject[]);

  return (
    <div>
      <Header activePage={activePage} setActivePage={setActivePage} />
      <MainComponent>
        <Switch>
          <Route exact path="/about" render={() => <AboutPage setActivePage={setActivePage} />} />
          <Route
            exact
            path="/details/:id"
            render={() => <DetailsPage setActivePage={setActivePage} />}
          />
          <Route
            exact
            path="/"
            render={() => (
              <HomePage setActivePage={setActivePage} setPageElements={setPageElements} />
            )}
          />
          <Route path="/*" render={() => <ErrorBlock />} />
        </Switch>
      </MainComponent>
    </div>
  );
};
export default App;
