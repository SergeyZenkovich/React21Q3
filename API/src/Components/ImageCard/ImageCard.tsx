import React from "react";
import styled from "styled-components";
import {FetchObject, ImageCardInterface} from "../../interfaces/ComponentsInterfaces";

import listImg from "../../assets/img/list.png";
import dateImg from "../../assets/img/calendar.png";
import likeImg from "../../assets/img/feedback.png";
import prodImg from "../../assets/img/prod.png";

import {StyledCard} from "./ImageCardStyling";

const ImageCard = ({id, color, likes, alt_description, user, urls}: FetchObject): JSX.Element => (
  <StyledCard>
    <img src={urls.regular} alt="" />
  </StyledCard>
);
export default ImageCard;
