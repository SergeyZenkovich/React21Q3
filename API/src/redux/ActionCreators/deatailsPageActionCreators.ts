import {FetchObject} from "../../interfaces/ComponentsInterfaces";
import {
  FetchActionsTypeDetails,
  LoadingDetailsAction,
  SetDetailsElementAction
} from "../../interfaces/reducersInterfaces";

export const LoadingActionCreator = (): LoadingDetailsAction => ({
  type: FetchActionsTypeDetails.TOGGLE_FETCHING
});
export const SetDetailsElementActionCreator = (element: FetchObject): SetDetailsElementAction => ({
  type: FetchActionsTypeDetails.SET_ELEMENT,
  element
});
