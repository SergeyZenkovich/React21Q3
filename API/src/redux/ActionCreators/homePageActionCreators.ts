import {FetchObject} from "src/interfaces/ComponentsInterfaces";
import {
  FetchActionsType,
  LoadingAction,
  SetElementsAction,
  SetTotalCountAction
} from "src/interfaces/reducersInterfaces";

export const toggleFetching = (): LoadingAction => ({
  type: FetchActionsType.TOGGLE_FETCHING
});
export const setElements = (elements: FetchObject[]): SetElementsAction => ({
  type: FetchActionsType.SET_ELEMENTS,
  elements
});
export const setTotalCount = (totalCount: number): SetTotalCountAction => ({
  type: FetchActionsType.SET_TOTAL_COUNT,
  totalCount
});
