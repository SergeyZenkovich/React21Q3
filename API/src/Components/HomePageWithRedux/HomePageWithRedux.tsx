import React, {useEffect} from "react";
import styled from "styled-components";
import {motion} from "framer-motion";
import {useDispatch} from "react-redux";
import {getFetchedDataThunkCreator} from "../../redux/Thunks/homePageThunks";
import {setRequestPageParam} from "../../redux/ActionCreators/homePageActionCreators";
import {HomePageStateInterface} from "../../interfaces/reducersInterfaces";
import Serach from "../SearchBar/SearchBar";
import ImageCard from "../ImageCard/ImageCard";
import Preloader from "../Preloader/Preloader";
import Pagination from "../Pagination/Pagination";
import Selector from "../SelectorForElementsPerPage/SelectorForElementsPerPage";
import ErrorBlock from "../ErrorComponent/ErrorComponent";

import {FetchObject} from "../../interfaces/ComponentsInterfaces";

import {
  StyledCardsBlockWithPagination,
  StyledCardsBlock,
  HomePageBlock
} from "../HomePage/HomePageStyling";

const HomePageWithRedux = ({
  homeState,
  setActivePage
}: {
  homeState: HomePageStateInterface;
  setActivePage: React.Dispatch<React.SetStateAction<string>>;
}): JSX.Element => {
  const {query, orient, orderBy, color, page, elementsOnPage} = homeState.requestParams;

  useEffect(() => {
    setActivePage("Home");
  }, []);
  const dispatch = useDispatch();
  const requestData = (pageNumber: number) => {
    dispatch(setRequestPageParam(pageNumber));
    dispatch(
      getFetchedDataThunkCreator({
        query,
        elementsOnPage,
        page: pageNumber,
        orient,
        color,
        orderBy
      })
    );
  };

  const ImagesCards =
    homeState.totalElementsCount === 0 ? (
      <ErrorBlock query={query} />
    ) : (
      homeState.elements.map((el: FetchObject) => (
        <ImageCard key={el.id} id={el.id} urls={el.urls} />
      ))
    );
  return (
    <HomePageBlock
      as={motion.div}
      initial={{opacity: 0}}
      animate={{opacity: 1}}
      exit={{opacity: 0}}
    >
      <Serach
        query={query}
        searchData={requestData}
        orderBy={orderBy}
        orient={orient}
        color={color}
      />
      <StyledCardsBlockWithPagination>
        <StyledCardsBlock>{homeState.isFetching ? <Preloader /> : ImagesCards}</StyledCardsBlock>
        {homeState.totalElementsCount >= 1 ? (
          <Pagination
            totalItemsCount={homeState.totalElementsCount}
            currentPage={page}
            portionSize={10}
            setFetchElems={requestData}
          />
        ) : null}
        {homeState.totalElementsCount >= 1 ? <Selector elementsOnPage={elementsOnPage} /> : null}
      </StyledCardsBlockWithPagination>
    </HomePageBlock>
  );
};
export default HomePageWithRedux;
