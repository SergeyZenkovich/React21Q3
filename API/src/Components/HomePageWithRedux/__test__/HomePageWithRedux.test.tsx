import React, {useState} from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import {render} from "@testing-library/react";
import {act} from "react-dom/test-utils";
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import {HashRouter} from "react-router-dom";
import Serach from "../../SearchBar/SearchBar";
import {FetchActionsType} from "../../../interfaces/reducersInterfaces";
import {setRequestPageParam} from "../../../redux/ActionCreators/homePageActionCreators";
import App from "../../../App";
import HomePageWithRedux from "../HomePageWithRedux";
import store from "../../../redux/store";
import {StyledCard} from "../../ImageCard/ImageCardStyling";

const testElement = {
  id: "testString",
  color: "#fff",
  alt_description: "test description",
  likes: 0,
  urls: {
    full: "fullTest",
    regular: "regularTest",
    small: "smallTest",
    thumb: "thumbTest"
  },
  user: {
    first_name: "testName",
    last_name: "testSurname",
    location: "testLocation",
    username: "testUserName",
    twitter_username: "testTwitter",
    instagram_username: "testIntagram",
    total_photos: 1,
    total_likes: 10
  }
};
const testElements = [testElement];

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const testObject = {
  totalElementsCount: -1,
  elements: [],
  isFetching: false,
  requestParams: {
    query: "",
    orderBy: "latest",
    orient: "landscape",
    color: "blue",
    page: 1,
    elementsOnPage: 10
  }
};
const testObjectWithRequest = {
  totalElementsCount: 1,
  elements: testElements,
  isFetching: false,
  requestParams: {
    query: "",
    orderBy: "latest",
    orient: "landscape",
    color: "blue",
    page: 1,
    elementsOnPage: 10
  }
};
const testNumber = 10;

it("renders without crashing", () => {
  const container = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <HomePageWithRedux homeState={testObject} setActivePage={() => {}}></HomePageWithRedux>
    </Provider>,
    container
  );
  expect(container.querySelector("input")).toBeTruthy();
});
it("renders without crashing and with image element block", () => {
  const container = document.createElement("div");
  ReactDOM.render(
    <HashRouter>
      <Provider store={store}>
        <HomePageWithRedux
          homeState={testObjectWithRequest}
          setActivePage={() => {}}
        ></HomePageWithRedux>
      </Provider>
    </HashRouter>,
    container
  );
  const innerBlockClassName = StyledCard.styledComponentId;
  const innerBlock = container.querySelector(`.${innerBlockClassName}`);
  expect(container.contains(innerBlock)).toBeTruthy();
  expect(container.querySelector("input")).toBeTruthy();
});

describe("HomePage Component", () => {
  const setFunction = jest.fn();
  let container;

  beforeEach(() => {
    container = renderer.create(
      <Provider store={store}>
        <HomePageWithRedux homeState={testObject} setActivePage={setFunction}></HomePageWithRedux>
      </Provider>
    );
  });
  it("matches shapshot", () => {
    const tree = container.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("useEffect Fires Once", () => {
    expect(setFunction).toBeCalledTimes(1);
  });
});
