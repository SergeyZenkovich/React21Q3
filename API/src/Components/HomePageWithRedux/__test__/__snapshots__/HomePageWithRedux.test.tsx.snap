// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component matches shapshot 1`] = `
<div
  className="sc-dIsUp FltwU"
  style={
    Object {
      "opacity": 0,
    }
  }
>
  <div
    className="sc-bdnxRM eMXXim"
  >
    <div
      className="sc-gtsrHT dXJXAG"
    >
      <input
        className="sc-hKFxyN eDeGEH"
        onChange={[Function]}
        placeholder="Iphone X...."
        value=""
      />
      <button
        className="sc-eCApnc hfjvEp"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
    <div
      className="sc-dlnjwi gRGhYz"
    >
      <select
        id="orderBy"
        name="orderBy"
        onChange={[Function]}
        value="latest"
      >
        <option
          value="relevant"
        >
          Relevant
        </option>
        <option
          value="latest"
        >
          Latest
        </option>
      </select>
      <select
        id="orientation"
        name="orientation"
        onChange={[Function]}
        value="landscape"
      >
        <option
          value="landscape"
        >
          Landscape
        </option>
        <option
          value="portrait"
        >
          Portrait
        </option>
        <option
          value="squarish"
        >
          Squarish
        </option>
      </select>
      <select
        id="color"
        name="color"
        onChange={[Function]}
        value="blue"
      >
        <option
          value="black_and_white"
        >
          Black and White
        </option>
        <option
          value="black"
        >
          Black
        </option>
        <option
          value="white"
        >
          White
        </option>
        <option
          value="yellow"
        >
          Yellow
        </option>
        <option
          value="orange"
        >
          Orange
        </option>
        <option
          value="red"
        >
          Red
        </option>
        <option
          value="purple"
        >
          Purple
        </option>
        <option
          value="magenta"
        >
          Magenta
        </option>
        <option
          value="green"
        >
          Green
        </option>
        <option
          value="teal"
        >
          Teal
        </option>
        <option
          value="blue"
        >
          Blue
        </option>
      </select>
    </div>
  </div>
  <div
    className="sc-iqAclL juntoK"
  >
    <div
      className="sc-crzoAE jdQjjQ"
    />
  </div>
</div>
`;
